certora/specs/YieldStreams.spec

```diff
@@ -0,0 +1,806 @@
+    require receiverPrincipal(bob, streamId) == _principal;
+    require _shares > 0;
+    require vault.balanceOf(alice) >= _shares;
+    require 1 <= parts;
+
+    mathint previewClaim = previewClaimYield(bob);
+    // Add X% profit to vault
+    require to_mathint(vault.balanceOf(currentContract)) == _shares + _principal / parts;
+
+    mathint claimed = claimYield(e, carol);
+
+    assert delta(claimed, previewClaim) <= 1 &&
+
+     vault.balanceOf(alice) == 0 &&
+     asset.balanceOf(bob) == 0 &&
+     delta(to_mathint(asset.balanceOf(carol)), claimed) <= 1;
+}
+
+/**
+ * @Rule Verify property for the `claimYield` function, with claims from all opened streams
+ * @Category High
+ * @Description  The `claimYield` function should verify the correct behavior of the claimYield function when the receiver (carol) claims the yield from multiple opened streams. It sets up a scenario where two yield streams are opened, one between alice and carol, and another between bob and carol, with specified principal amounts. The rule then generates yield by adding a percentage of the principals to the vault's balance. Finally, it invokes the claimYield function with carol as the receiver and asserts that the claimed yield is correctly calculated and transferred to carol's asset balance, while ensuring that the contract's state is updated accordingly.
+ */
+
+rule verify_claimYield_claimsFromAllOpenedStreams(address alice, address carol, address bob, uint256 alicesPrincipal, uint256 bobsPrincipal, uint8 multiple, uint8 desloc) {
+    env e;
+    require carol == e.msg.sender && alice != carol && alice != bob && bob != carol;
+    require to_mathint(alicesPrincipal) == multiple * (10 ^ 18);
+    uint256 _aliceShares = vault.convertToShares(alicesPrincipal);
+    uint256 aliceStreamId = nextStreamId();
+    require streamIdToReceiver(aliceStreamId) == carol;
+    require receiverPrincipal(carol, aliceStreamId) == alicesPrincipal;
+    require vault.balanceOf(alice) >= _aliceShares;
+
+    require to_mathint(bobsPrincipal) == (multiple + desloc) * (10 ^ 18);
+    uint256 _bobShares = vault.convertToShares(bobsPrincipal);
+    uint256 bobStreamId = aliceStreamId;
+    require streamIdToReceiver(bobStreamId) == carol;
+    require to_mathint(receiverTotalPrincipal(carol)) >= alicesPrincipal + bobsPrincipal;
+    require receiverPrincipal(carol, bobStreamId) == bobsPrincipal;
+    require vault.balanceOf(bob) >= _bobShares;
+
+
+    // add X% profit to vault
+    require 1 <= desloc;
+    require to_mathint(receiverTotalShares(carol)) >= _aliceShares + _bobShares + (multiple + desloc) * (10 ^ 18);
+    require to_mathint(vault.balanceOf(currentContract)) >= _aliceShares + _bobShares + (multiple + desloc) * (10 ^ 18);
+    
+    assert delta(to_mathint(previewClaimYield(carol)), alicesPrincipal + bobsPrincipal) <= 1;
+
+    uint256 claimed = claimYield(e, carol);
+
+    assert to_mathint(claimed) == alicesPrincipal + bobsPrincipal &&
```

Comment:
perhaps multiple asserts is easier to read than `&&`, also it would be easier to investigate assertion failure if each assert is about one single postcondition

Reply:
Separated